Trailing newlines get swallowed

The new pretty printer does not output any newlines at the end of a file.

**Description**

The `CanonicalPrinter` defined in `ttcn3/v2/printer` should output a single newline character (`\n`) for every one or more line break (e.g. `\r\n\r\n`) in the input source.  
The current implementation however does not output any newlines characters for the last token.

**To Reproduce**
I've prepared a test [here](https://github.com/nokia/ntt/blob/master/ttcn3/v2/printer/printer_test.go#L19). It is currently skiped, to execute this test:
1. Remove `skip` from linked test.
2. Run test.

```
$ go test -run 'TestSimpleFormatter/TestSimpleFormatter#02'
--- FAIL: TestSimpleFormatter (0.00s)
    --- FAIL: TestSimpleFormatter/TestSimpleFormatter#02 (0.00s)
        printer_test.go:76: 
            	Error Trace:	printer_test.go:76
            	Error:      	Not equal: 
            	            	expected: "foo;\n"
            	            	actual  : "foo;"
            	            	
            	            	Diff:
            	            	--- Expected
            	            	+++ Actual
            	            	@@ -1,2 +1 @@
            	            	 foo;
            	            	-
            	Test:       	TestSimpleFormatter/TestSimpleFormatter#02
FAIL
exit status 1
FAIL	github.com/nokia/ntt/ttcn3/v2/printer	0.003s

```

**Expected behavior**

A trailing newline should not disappear. Examples:

        input: "foo;\n\n\n\n",  want: "foo;\n"
        input: "//foo", want: "//foo"
        input: "//foo\n", want: "//foo\n"

Alternatively I think it would be okay if the last token always has a trailing newline. Examples:

        input: "foo;\n\n\n\n",  want: "foo;\n"
        input: "//foo", want: "//foo\n"
        input: "//foo\n", want: "//foo\n"

--- a/ttcn3/v2/printer/example_test.go
+++ b/ttcn3/v2/printer/example_test.go
@@ -1,0 +1,31 @@
+package printer_test
+
+import (
+	"os"
+
+	"github.com/nokia/ntt/ttcn3/v2/printer"
+)
+
+func ExampleFprint() {
+	input := "{\n x:=1, // Comment\ny := 234, // Comment 2\n}"
+	if err := printer.Fprint(os.Stdout, input); err != nil {
+		panic(err)
+	}
+	// Output:
+	// {
+	// 	x := 1,   // Comment
+	// 	y := 234, // Comment 2
+	// }
+}
+
+func ExampleCanonicalPrinter() {
+	input := "{\n x:=1, // Comment\ny := 234, // Comment 2\n}"
+	p := printer.NewCanonicalPrinter(os.Stdout)
+	p.Indent = "  "
+	p.Fprint(input)
+	// Output:
+	// {
+	//   x := 1,   // Comment
+	//   y := 234, // Comment 2
+	// }
+}

--- a/ttcn3/v2/printer/printer.go
+++ b/ttcn3/v2/printer/printer.go
@@ -5,133 +5,233 @@ import (
 	"fmt"
 	"io"
 	"strings"
+	"text/tabwriter"
 
 	"github.com/nokia/ntt/ttcn3/v2/syntax"
+)
+
+type whiteSpace byte
+
+const (
+	ignore   = whiteSpace(0)
+	blank    = whiteSpace(' ')
+	newline  = whiteSpace('\n')
+	cell     = whiteSpace('|')
+	indent   = whiteSpace('>')
+	unindent = whiteSpace('<')
 )
 
 // Fprint formats src in canonical TTCN-3 style and writes the result to w or
 // returns an (I/O or syntax) error. src is expected to be syntactically
 // correct TTCN-3 source text.
 func Fprint(w io.Writer, src interface{}) error {
-	p := &CanonicalPrinter{
-		Indent: "\t",
-		w:      w,
-	}
-	return p.Fprint(src)
+	return NewCanonicalPrinter(w).Fprint(src)
 }
 
 // CanonicalPrinter is a simple formatter that only fixes indentation and
 // various whitespace issues.
 type CanonicalPrinter struct {
-	Indent     string // Indentation string; default is "\t"
-	w          io.Writer
-	lastPos    syntax.Position
-	indent     int
-	needIndent bool
+	Indent           string // Indentation string; default is "\t"
+	w                *tabwriter.Writer
+	lastPos, currPos syntax.Position
+	indent           int
+	whiteBuf         string
+	firstToken       bool
+}
+
+// NewCanonicalPrinter returns a new printer that formats source code.
+func NewCanonicalPrinter(w io.Writer) *CanonicalPrinter {
+	return &CanonicalPrinter{
+		indent: 0,
+		Indent: "\t",
+		w: tabwriter.NewWriter(w, 0, 8, 1, ' ',
+			tabwriter.TabIndent|tabwriter.StripEscape|tabwriter.DiscardEmptyColumns),
+	}
 }
 
 func (p *CanonicalPrinter) Fprint(v interface{}) error {
 
-	var (
-		n syntax.Node
-
-		// The simple formatting rules do not need any context information
-		// so far. This allows us to use the tokeniser and release initial
-		// formatting experiments even before the parser is ready.
-		parse = syntax.Tokenize
-	)
-
-	switch v := v.(type) {
-	case []byte:
-		n = parse(v)
-	case string:
-		n = parse([]byte(v))
-	case io.Reader:
-		b, err := io.ReadAll(v)
-		if err != nil {
-			return err
-		}
-		n = parse(b)
-	default:
-		return fmt.Errorf("printer: unsupported type %T", v)
+	b, err := toBytes(v)
+	if err != nil {
+		return err
 	}
+
+	// The simple formatting rules do not need much context information.
+	// This allows us to use the tokeniser and release initial
+	// formatting experiments even before the parser is ready.
+	n := syntax.Tokenize(b)
 
 	// Only pretty print if there are no syntax errors.
 	if n.Err() != nil {
 		return n.Err()
 	}
 
-	// Prime the position tracker with the first token.
-	if tok := n.FirstToken(); tok != syntax.Nil {
-		s := tok.Span()
-		p.lastPos = s.End
-	}
-
-	return p.node(n)
+	return p.tree(n)
 }
 
-func (p *CanonicalPrinter) node(n syntax.Node) error {
+// tree prints the syntax tree by interspersing the token stream with spacing
+// information.
+func (p *CanonicalPrinter) tree(n syntax.Node) error {
+
+	// Prime the position tracker with the first token to avoid printing
+	// white-spaces before the first token.
+	p.firstToken = true
+
 	n.Inspect(func(n syntax.Node) bool {
-		if n == syntax.Nil || !n.IsToken() {
+
+		// We figure out spacing by looking at the token stream only.
+		// Because the parser for creating a syntax tree is not yet
+		// implemented.
+		if !n.IsToken() {
 			return true
 		}
 
-		// Handle user defined white space
+		// Incorporate user-defined line breaks and token separators
+		// into output stream.
 		currPos := n.Span()
 		switch {
 		case currPos.Begin.Line > p.lastPos.Line:
-			p.print("\n")
+			p.print(newline)
 			if currPos.Begin.Line-p.lastPos.Line > 1 {
-				p.print("\n")
+				p.print(newline)
 			}
 		case currPos.Begin.Column > p.lastPos.Column:
-			p.print(" ")
+			p.print(blank)
 		}
 		p.lastPos = currPos.End
 
-		// Handle indent.
-		//
-		// Indentation is usually handled by non-terminal nodes. But the pretty
-		// printer rules are very simple and we can handle it here.
-		//
-		// The rule is: Increment indentation after every opening brace and
-		// decrement before every closing brace.
-		switch s := n.Text(); {
-		case s == "{", s == "[", s == "(":
-			p.print(s)
-			p.indent++
-		case s == "}", s == "]", s == ")":
-			p.indent--
-			p.print(s)
-		case strings.HasPrefix(s, "/*"):
-			lines := strings.Split(s, "\n")
-			for i, line := range lines {
-				if i == 0 {
-					p.print(line)
-				} else {
-					p.print("\n")
-					p.print(" ")
-					p.print(line)
-				}
+		switch k, s := n.Kind(), n.Text(); {
+
+		// Increment indentation after opening {, [, (.
+		case k == syntax.LeftBrace, k == syntax.LeftBracket, k == syntax.LeftParen:
+			p.print(s, indent)
+
+		// Decrement indentation before closing }, ], ).
+		case k == syntax.RightBrace, k == syntax.RightBracket, k == syntax.RightParen:
+			p.print(unindent, s)
+
+		// Align assignments.
+		case k == syntax.Assign:
+			p.print(cell, ":=", blank)
+
+		// Every line of a comment has to be indented individually.
+		case k == syntax.Comment:
+			p.print(cell)
+			p.comment(strings.Split(strings.TrimSpace(s), "\n"))
+			if strings.HasSuffix(s, "\n") {
+				p.print(newline)
 			}
+
+		// Only literals may contain newlines and \t and need to be quoted.
+		case n.Kind().IsLiteral():
+			p.print(quote(s))
+
+		// All other tokens are printed as is.
 		default:
 			p.print(s)
 		}
-
 		return true
 	})
-	return nil
+
+	// Terminate the last line with a newline.
+	if !p.firstToken {
+		fmt.Fprint(p.w, "\n")
+	}
+
+	return p.w.Flush()
 }
 
-func (p *CanonicalPrinter) print(s string) {
-	if p.needIndent {
+// comment prints a comment line by line. It removes white-space prefixes from
+// multi-line comments, so they can be properly indented.
+func (p *CanonicalPrinter) comment(lines []string) {
+	line, lines := lines[0], lines[1:]
+	p.print(quote(line))
+
+	if len(lines) == 0 {
+		return
+	}
+
+	prefix := ""
+	if l := lines[len(lines)-1]; strings.HasSuffix(l, "*/") {
+		l = l[0 : len(l)-2]
+		if strings.TrimSpace(l) == "" {
+			prefix = l
+		}
+	}
+
+	for _, line := range lines {
+		p.print(newline)
+		p.print(" ", quote(strings.TrimPrefix(line, prefix)))
+	}
+}
+
+func (p *CanonicalPrinter) print(args ...interface{}) {
+	for _, arg := range args {
+		switch arg := arg.(type) {
+		case whiteSpace:
+
+			switch arg {
+			case ignore:
+				continue
+			case indent:
+				p.indent++
+			case unindent:
+				p.indent--
+			case blank:
+				if p.whiteBuf == "" {
+					p.whiteBuf = " "
+				}
+			case cell:
+				if p.whiteBuf == " " {
+					p.whiteBuf = ""
+				}
+				if strings.Count(p.whiteBuf, "\t") == len(p.whiteBuf) {
+					p.whiteBuf += "\t"
+				}
+			case newline:
+				if strings.Count(p.whiteBuf, "\n") != len(p.whiteBuf) {
+					p.whiteBuf = ""
+				}
+				p.whiteBuf += "\n"
+			}
+		default:
+			if p.whiteBuf != "" {
+				p.printSpace()
+				p.whiteBuf = ""
+			}
+			fmt.Fprint(p.w, arg)
+		}
+
+	}
+}
+
+// printSpace prints the white-space buffer and indentation.
+func (p *CanonicalPrinter) printSpace() {
+	if p.firstToken {
+		p.firstToken = false
+		return
+	}
+	fmt.Fprint(p.w, p.whiteBuf)
+	if strings.HasSuffix(p.whiteBuf, "\n") {
 		for i := 0; i < p.indent; i++ {
 			fmt.Fprint(p.w, p.Indent)
 		}
-		p.needIndent = false
 	}
-	fmt.Fprint(p.w, s)
-	if strings.HasSuffix(s, "\n") {
-		p.needIndent = true
+}
+
+func toBytes(v interface{}) ([]byte, error) {
+	switch v := v.(type) {
+	case []byte:
+		return v, nil
+	case string:
+		return []byte(v), nil
+	case io.Reader:
+		return io.ReadAll(v)
+	default:
+		return nil, fmt.Errorf("printer: unsupported type %T", v)
 	}
+}
+
+func quote(s string) string {
+	return fmt.Sprintf("\xff%s\xff", s)
 }

--- a/ttcn3/v2/printer/printer_test.go
+++ b/ttcn3/v2/printer/printer_test.go
@@ -15,51 +15,56 @@ func TestSimpleFormatter(t *testing.T) {
 		skip  bool
 	}{
 		{input: "", want: ""},
-		{input: "foo;", want: "foo;"},
-		{input: "foo;\n\n", want: "foo;\n", skip: true}, // trailing newline does not work yet.
-		{input: "//foo", want: "//foo"},
+		{input: "foo;", want: "foo;\n"},
+		{input: "foo;\n\n", want: "foo;\n"},
+		{input: "//foo", want: "//foo\n"},
 		{input: "//foo\n", want: "//foo\n"},
 
 		// Remove leading whitespace
-		{input: "    leading;", want: "leading;"},
+		{input: "    leading;", want: "leading;\n"},
 
 		// Remove trailing whitespace
-		{input: "trailing;   ", want: "trailing;"},
+		{input: "trailing;   ", want: "trailing;\n"},
 
 		// At max one blank between tokens.
-		{input: "import from   all", want: "import from all"},
+		{input: "import from   all", want: "import from all\n"},
 
 		// Convert all other whitespace to blanks.
-		{input: "import \tfrom\tall", want: "import from all"},
+		{input: "import \tfrom\tall", want: "import from all\n"},
 
 		// Replace line breaks with \n
-		{input: "foo;\r\nbar;", want: "foo;\nbar;"},
-		{input: "foo;\rbar;", want: "foo; bar;"}, // \r is not a line break
-		{input: "foo;\n\rbar;", want: "foo;\nbar;"},
-		{input: "foo;\vbar;", want: "foo;\nbar;"},
-		{input: "foo;\fbar;", want: "foo;\nbar;"},
+		{input: "foo;\r\nbar;", want: "foo;\nbar;\n"},
+		{input: "foo;\rbar;", want: "foo; bar;\n"}, // \r is not a line break
+		{input: "foo;\n\rbar;", want: "foo;\nbar;\n"},
+		{input: "foo;\vbar;", want: "foo;\nbar;\n"},
+		{input: "foo;\fbar;", want: "foo;\nbar;\n"},
 
 		// Keep at most one newline
-		{input: "foo;\r\n\r\nbar;", want: "foo;\n\nbar;"},
+		{input: "foo;\r\n\r\nbar;", want: "foo;\n\nbar;\n"},
 
 		// User defined spaces
-		{input: "control{}", want: "control{}"},
-		{input: "control {}", want: "control {}"},
-		{input: "control {} // Foo", want: "control {} // Foo"},
-		{input: "control  {}  ", want: "control {}"},
-		{input: "control \n{}", want: "control\n{}"},
-		{input: "control\n {}", want: "control\n{}"},
+		{input: "control{}", want: "control{}\n"},
+		{input: "control {}", want: "control {}\n"},
+		{input: "control {} // Foo", want: "control {} // Foo\n"},
+		{input: "control  {}  ", want: "control {}\n"},
+		{input: "control \n{}", want: "control\n{}\n"},
+		{input: "control\n {}", want: "control\n{}\n"},
 
 		// Verify that {, [ and ( increment the indentation level
-		{input: "{foo", want: "{foo"},
-		{input: "{\nfoo", want: "{\n\tfoo"},
-		{input: "{\n foo", want: "{\n\tfoo"},
-		{input: "{\nfoo}\nbar", want: "{\n\tfoo}\nbar"},
-		{input: "{\n[\n(\n1,2\n)\n]\n}", want: "{\n\t[\n\t\t(\n\t\t\t1,2\n\t\t)\n\t]\n}"},
+		{input: "{foo", want: "{foo\n"},
+		{input: "{\nfoo", want: "{\n\tfoo\n"},
+		{input: "{\n foo", want: "{\n\tfoo\n"},
+		{input: "{\nfoo}\nbar", want: "{\n\tfoo}\nbar\n"},
+		{input: "{\n[\n(\n1,2\n)\n]\n}", want: "{\n\t[\n\t\t(\n\t\t\t1,2\n\t\t)\n\t]\n}\n"},
 
 		// Verify that tokens with newlines have correct indentation
-		{input: "{// Foo\nBar", want: "{// Foo\n\tBar"},               //  Bar must be indented.
-		{input: "{\n/*\n* foo\n*/", want: "{\n\t/*\n\t * foo\n\t */"}, //  Comment must be indented, with one extra space.
+		{input: "{// Foo\nBar", want: "{ // Foo\n\tBar\n"},              //  Bar must be indented.
+		{input: "{\n/*\n* foo\n*/", want: "{\n\t/*\n\t * foo\n\t */\n"}, //  Comment must be indented, with one extra space.
+
+		// Verify that comments and := are aligned.
+		{input: "{x := 1,\nx2:= 123}", want: "{x := 1,\n\tx2 := 123}\n"},
+		{input: "{\nx := 1,\nx2:= 123}", want: "{\n\tx  := 1,\n\tx2 := 123}\n"},
+		{input: "{\nx := 1, // a\nx2:= 123 /* b */}", want: "{\n\tx  := 1,  // a\n\tx2 := 123 /* b */}\n"},
 	}
 
 	for _, test := range tests {

